File: libs/shared-types/src/lib/catalog.ts
/**
 * Catalog transport types shared between apps/api, web, tma, admin.
 */

import type { PickupPointType } from './shared-types';

export type StoreStatus = 'OPEN' | 'CLOSED' | 'OVERLOADED';
export type StoreFulfillment = 'TAKEAWAY' | 'DINE_IN' | 'DRIVE_THRU' | 'DELIVERY';
export type VariationType = 'SIZE' | 'TEMPERATURE' | 'MILK' | 'CUP';
export type DietTag = 'VEGAN' | 'VEGETARIAN' | 'GLUTEN_FREE' | 'LACTOSE_FREE' | 'DECAF' | 'SUGAR_FREE';

export type { PickupPointType };

export interface StoreListItem {
  id: string;
  brandId: string;
  slug: string;
  name: string;
  addressLine: string;
  city: string;
  country: string;
  latitude: number;
  longitude: number;
  status: StoreStatus;
  fulfillmentTypes: StoreFulfillment[];
  pickupPointType: PickupPointType;
  busyMeter: number;
  currentEtaSeconds: number;
  /**
   * An ASAP order placed now would be accepted: not closed, and within the
   * store's working hours once the current ETA has passed. When false only
   * scheduled pickup works.
   */
  openNow: boolean;
  /** Sales tax in basis points: 500 = 5%, 2000 = 20%. */
  taxRateBps: number;
  /** True when the listed prices already include the tax. */
  taxIncludedInPrice: boolean;
  currency: string;
  /** IANA zone. Pickup and opening times are shown on the store's clock. */
  timezone: string;
  heroImageUrl: string | null;
  distanceMeters: number | null;
}

/**
 * One 15-minute handover window. Full slots still come back, marked
 * unavailable, so checkout can grey them out — a missing time reads as a
 * bug, a greyed-out one reads as a busy morning.
 */
export interface PickupSlot {
  startsAt: string;
  endsAt: string;
  taken: number;
  capacity: number;
  available: boolean;
}

export interface StoreWorkingHour {
  weekday: number;
  opensAt: number;
  closesAt: number;
  isClosed: boolean;
}

export interface BrandTheme {
  id: string;
  slug: string;
  name: string;
  logoUrl: string | null;
  /** CSS-variable overrides applied on top of the Telegram theme. */
  themeOverrides: Record<string, string> | null;
}

export interface StoreDetail extends StoreListItem {
  phone: string | null;
  email: string | null;
  minOrderCents: number;
  galleryUrls: string[];
  workingHours: StoreWorkingHour[];
  brand: BrandTheme;
}

export interface Variation {
  id: string;
  type: VariationType;
  name: string;
  priceDeltaCents: number;
  prepTimeDeltaSeconds: number;
  sortOrder: number;
  isDefault: boolean;
}

export interface Modifier {
  id: string;
  slug: string;
  name: string;
  priceDeltaCents: number;
  prepTimeDeltaSeconds: number;
  minCount: number;
  maxCount: number;
  sortOrder: number;
}

export interface ProductSummary {
  id: string;
  categoryId: string;
  slug: string;
  name: string;
  description: string | null;
  basePriceCents: number;
  prepTimeSeconds: number;
  caffeineLevel: number | null;
  calories: number | null;
  proteinsGrams: number | null;
  fatsGrams: number | null;
  carbsGrams: number | null;
  allergens: string[];
  dietTags: DietTag[];
  imageUrls: string[];
  sortOrder: number;
  onStopList?: boolean;
}

export interface ProductDetail extends ProductSummary {
  /** Brand that owns this product — only a store of this brand can make it. */
  brandId: string;
  variations: Variation[];
  modifiers: Modifier[];
}

export interface CategoryWithProducts {
  id: string;
  slug: string;
  name: string;
  description: string | null;
  iconUrl: string | null;
  sortOrder: number;
  availableFrom: number | null;
  availableTo: number | null;
  products: ProductSummary[];
}

export interface StoreMenu {
  storeId: string;
  storeSlug: string;
  categories: CategoryWithProducts[];
}

export interface ListStoresQuery {
  lat?: number;
  lng?: number;
  radius?: number;
}

File: libs/api-client-dart/lib/src/models/catalog.dart
import 'package:json_annotation/json_annotation.dart';

part 'catalog.g.dart';

@JsonEnum(alwaysCreate: true)
enum StoreStatus {
  @JsonValue('OPEN')
  open,
  @JsonValue('OVERLOADED')
  overloaded,
  @JsonValue('CLOSED')
  closed,
  unknown,
}

@JsonEnum(alwaysCreate: true)
enum VariationType {
  @JsonValue('SIZE')
  size,
  @JsonValue('TEMPERATURE')
  temperature,
  @JsonValue('MILK')
  milk,
  @JsonValue('CUP')
  cup,
  unknown,
}

@JsonSerializable(explicitToJson: true)
class Store {
  const Store({
    required this.id,
    required this.brandId,
    required this.slug,
    required this.name,
    required this.addressLine,
    required this.city,
    required this.country,
    required this.latitude,
    required this.longitude,
    required this.status,
    required this.fulfillmentTypes,
    required this.pickupPointType,
    required this.busyMeter,
    required this.currentEtaSeconds,
    required this.taxRateBps,
    required this.taxIncludedInPrice,
    required this.currency,
    this.openNow,
    this.heroImageUrl,
    this.distanceMeters,
  });

  factory Store.fromJson(Map<String, dynamic> json) => _$StoreFromJson(json);

  Map<String, dynamic> toJson() => _$StoreToJson(this);

  final String id;
  final String brandId;
  final String slug;
  final String name;
  final String addressLine;
  final String city;
  final String country;
  final double latitude;
  final double longitude;
  @JsonKey(unknownEnumValue: StoreStatus.unknown)
  final StoreStatus status;
  @JsonKey(defaultValue: <String>[])
  final List<String> fulfillmentTypes;
  final String pickupPointType;
  final int busyMeter;
  final int currentEtaSeconds;

  /// Whether an ASAP order placed now would be accepted — the manual switch
  /// plus working hours, computed by the API in the store's timezone. Null
  /// from API versions that predate it.
  final bool? openNow;
  @JsonKey(defaultValue: 0)
  final int taxRateBps;
  @JsonKey(defaultValue: true)
  final bool taxIncludedInPrice;
  final String currency;
  final String? heroImageUrl;
  final double? distanceMeters;

  bool get hasLocation => latitude != 0 || longitude != 0;

  /// Takes ASAP orders right now. Falls back to the manual switch alone when
  /// the API does not send [openNow].
  bool get isOpen => status != StoreStatus.closed && (openNow ?? true);

  /// [status] with working hours applied: a store switched on but outside its
  /// hours reads as closed.
  StoreStatus get effectiveStatus => isOpen ? status : StoreStatus.closed;
  bool get supportsDelivery => fulfillmentTypes.contains('DELIVERY');
  String get fullAddress => [addressLine, city].where((p) => p.trim().isNotEmpty).join(', ');
}

@JsonSerializable(explicitToJson: true)
class StoreWorkingHour {
  const StoreWorkingHour({
    required this.weekday,
    required this.opensAt,
    required this.closesAt,
    required this.isClosed,
  });

  factory StoreWorkingHour.fromJson(Map<String, dynamic> json) => _$StoreWorkingHourFromJson(json);

  Map<String, dynamic> toJson() => _$StoreWorkingHourToJson(this);

  /// 0 = Sunday … 6 = Saturday.
  final int weekday;

  /// Minutes since local midnight.
  final int opensAt;
  final int closesAt;
  final bool isClosed;
}

@JsonSerializable(explicitToJson: true)
class BrandInfo {
  const BrandInfo({required this.id, required this.slug, required this.name, this.logoUrl, this.themeOverrides});

  factory BrandInfo.fromJson(Map<String, dynamic> json) => _$BrandInfoFromJson(json);

  Map<String, dynamic> toJson() => _$BrandInfoToJson(this);

  final String id;
  final String slug;
  final String name;
  final String? logoUrl;
  final Map<String, dynamic>? themeOverrides;
}

@JsonSerializable(explicitToJson: true)
class StoreDetail extends Store {
  const StoreDetail({
    required super.id,
    required super.brandId,
    required super.slug,
    required super.name,
    required super.addressLine,
    required super.city,
    required super.country,
    required super.latitude,
    required super.longitude,
    required super.status,
    required super.fulfillmentTypes,
    required super.pickupPointType,
    required super.busyMeter,
    required super.currentEtaSeconds,
    required super.taxRateBps,
    required super.taxIncludedInPrice,
    required super.currency,
    required this.timezone,
    required this.minOrderCents,
    required this.galleryUrls,
    required this.workingHours,
    required this.brand,
    super.openNow,
    super.heroImageUrl,
    super.distanceMeters,
    this.phone,
    this.email,
  });

  factory StoreDetail.fromJson(Map<String, dynamic> json) => _$StoreDetailFromJson(json);

  @override
  Map<String, dynamic> toJson() => _$StoreDetailToJson(this);

  final String timezone;
  final String? phone;
  final String? email;
  @JsonKey(defaultValue: 0)
  final int minOrderCents;
  @JsonKey(defaultValue: <String>[])
  final List<String> galleryUrls;
  @JsonKey(defaultValue: <StoreWorkingHour>[])
  final List<StoreWorkingHour> workingHours;
  final BrandInfo brand;
}

@JsonSerializable(explicitToJson: true)
class Variation {
  const Variation({
    required this.id,
    required this.type,
    required this.name,
    required this.priceDeltaCents,
    required this.prepTimeDeltaSeconds,
    required this.sortOrder,
    required this.isDefault,
  });

  factory Variation.fromJson(Map<String, dynamic> json) => _$VariationFromJson(json);

  Map<String, dynamic> toJson() => _$VariationToJson(this);

  final String id;
  @JsonKey(unknownEnumValue: VariationType.unknown)
  final VariationType type;
  final String name;
  final int priceDeltaCents;
  @JsonKey(defaultValue: 0)
  final int prepTimeDeltaSeconds;
  @JsonKey(defaultValue: 0)
  final int sortOrder;
  @JsonKey(defaultValue: false)
  final bool isDefault;
}

@JsonSerializable(explicitToJson: true)
class Modifier {
  const Modifier({
    required this.id,
    required this.slug,
    required this.name,
    required this.priceDeltaCents,
    required this.prepTimeDeltaSeconds,
    required this.minCount,
    required this.maxCount,
    required this.sortOrder,
  });

  factory Modifier.fromJson(Map<String, dynamic> json) => _$ModifierFromJson(json);

  Map<String, dynamic> toJson() => _$ModifierToJson(this);

  final String id;
  final String slug;
  final String name;
  final int priceDeltaCents;
  @JsonKey(defaultValue: 0)
  final int prepTimeDeltaSeconds;
  @JsonKey(defaultValue: 0)
  final int minCount;
  @JsonKey(defaultValue: 1)
  final int maxCount;
  @JsonKey(defaultValue: 0)
  final int sortOrder;
}

@JsonSerializable(explicitToJson: true)
class Product {
  const Product({
    required this.id,
    required this.categoryId,
    required this.slug,
    required this.name,
    required this.basePriceCents,
    required this.prepTimeSeconds,
    required this.allergens,
    required this.dietTags,
    required this.imageUrls,
    required this.sortOrder,
    this.description,
    this.caffeineLevel,
    this.calories,
    this.proteinsGrams,
    this.fatsGrams,
    this.carbsGrams,
    this.onStopList = false,
  });

  factory Product.fromJson(Map<String, dynamic> json) => _$ProductFromJson(json);

  Map<String, dynamic> toJson() => _$ProductToJson(this);

  final String id;
  final String categoryId;
  final String slug;
  final String name;
  final String? description;
  final int basePriceCents;
  @JsonKey(defaultValue: 0)
  final int prepTimeSeconds;
  final int? caffeineLevel;
  final int? calories;
  final double? proteinsGrams;
  final double? fatsGrams;
  final double? carbsGrams;
  @JsonKey(defaultValue: <String>[])
  final List<String> allergens;
  @JsonKey(defaultValue: <String>[])
  final List<String> dietTags;
  @JsonKey(defaultValue: <String>[])
  final List<String> imageUrls;
  @JsonKey(defaultValue: 0)
  final int sortOrder;
  @JsonKey(defaultValue: false)
  final bool onStopList;

  String? get imageUrl => imageUrls.isEmpty ? null : imageUrls.first;
  bool get hasNutrition => proteinsGrams != null || fatsGrams != null || carbsGrams != null;
}

@JsonSerializable(explicitToJson: true)
class ProductDetail extends Product {
  const ProductDetail({
    required super.id,
    required super.categoryId,
    required super.slug,
    required super.name,
    required super.basePriceCents,
    required super.prepTimeSeconds,
    required super.allergens,
    required super.dietTags,
    required super.imageUrls,
    required super.sortOrder,
    required this.brandId,
    required this.variations,
    required this.modifiers,
    super.description,
    super.caffeineLevel,
    super.calories,
    super.proteinsGrams,
    super.fatsGrams,
    super.carbsGrams,
    super.onStopList,
  });

  factory ProductDetail.fromJson(Map<String, dynamic> json) => _$ProductDetailFromJson(json);

  @override
  Map<String, dynamic> toJson() => _$ProductDetailToJson(this);

  final String brandId;
  @JsonKey(defaultValue: <Variation>[])
  final List<Variation> variations;
  @JsonKey(defaultValue: <Modifier>[])
  final List<Modifier> modifiers;
}

@JsonSerializable(explicitToJson: true)
class MenuCategory {
  const MenuCategory({
    required this.id,
    required this.slug,
    required this.name,
    required this.sortOrder,
    required this.products,
    this.description,
    this.iconUrl,
    this.availableFrom,
    this.availableTo,
  });

  factory MenuCategory.fromJson(Map<String, dynamic> json) => _$MenuCategoryFromJson(json);

  Map<String, dynamic> toJson() => _$MenuCategoryToJson(this);

  final String id;
  final String slug;
  final String name;
  final String? description;
  final String? iconUrl;
  @JsonKey(defaultValue: 0)
  final int sortOrder;
  final int? availableFrom;
  final int? availableTo;
  @JsonKey(defaultValue: <Product>[])
  final List<Product> products;
}

@JsonSerializable(explicitToJson: true)
class Menu {
  const Menu({required this.storeId, required this.storeSlug, required this.categories});

  factory Menu.fromJson(Map<String, dynamic> json) => _$MenuFromJson(json);

  Map<String, dynamic> toJson() => _$MenuToJson(this);

  final String storeId;
  final String storeSlug;
  @JsonKey(defaultValue: <MenuCategory>[])
  final List<MenuCategory> categories;

  /// Categories that actually have something to show.
  List<MenuCategory> get visibleCategories => categories.where((c) => c.products.isNotEmpty).toList();

  Iterable<Product> get allProducts => categories.expand((c) => c.products);
}

@JsonSerializable(explicitToJson: true)
class PickupSlot {
  const PickupSlot({
    required this.startsAt,
    required this.endsAt,
    required this.taken,
    required this.capacity,
    required this.available,
  });

  factory PickupSlot.fromJson(Map<String, dynamic> json) => _$PickupSlotFromJson(json);

  Map<String, dynamic> toJson() => _$PickupSlotToJson(this);

  final DateTime startsAt;
  final DateTime endsAt;
  final int taken;
  final int capacity;
  final bool available;
}
